keanu-project/src/main/java/io/improbable/keanu/distributions/discrete/Multinomial.java

```diff
@@ -1,116 +1,96 @@
 package io.improbable.keanu.distributions.discrete;
 
 import com.google.common.base.Preconditions;
+import com.google.common.primitives.Ints;
 import io.improbable.keanu.KeanuRandom;
 import io.improbable.keanu.distributions.DiscreteDistribution;
-import io.improbable.keanu.tensor.Tensor;
 import io.improbable.keanu.tensor.TensorShape;
-import io.improbable.keanu.tensor.TensorShapeValidation;
 import io.improbable.keanu.tensor.dbl.DoubleTensor;
 import io.improbable.keanu.tensor.intgr.IntegerTensor;
-import io.improbable.keanu.tensor.validate.DebugTensorValidator;
-import io.improbable.keanu.tensor.validate.TensorValidator;
-import org.apache.commons.lang3.ArrayUtils;
 
 import java.util.Arrays;
-import java.util.List;
-import java.util.stream.Collectors;
 
-import static io.improbable.keanu.tensor.TensorShapeValidation.checkTensorsMatchNonLengthOneShapeOrAreLengthOne;
+import static io.improbable.keanu.tensor.TensorShapeValidation.isBroadcastable;
 
 
 public class Multinomial implements DiscreteDistribution {
 
-    public static final DebugTensorValidator CATEGORY_PROBABILITIES_CANNOT_BE_ZERO = new DebugTensorValidator<>(TensorValidator.ZERO_CATCHER);
     private final IntegerTensor n;
     private final DoubleTensor p;
-    private final long numCategories;
+    private final int k;
+    private boolean validationEnabled;
 
     public static Multinomial withParameters(IntegerTensor n, DoubleTensor p) {
-        return new Multinomial(n, p);
+        return new Multinomial(n, p, false);
+    }
+
+    public static Multinomial withParameters(IntegerTensor n, DoubleTensor p, boolean validationEnabled) {
+        return new Multinomial(n, p, validationEnabled);
     }
 
     /**
      * @param n The number of draws from the variable
      * @param p The probability of observing each of the k values (which sum to 1)
-     *          p is a Tensor whose first dimension must be of size k
+     *          p is a Tensor whose last dimension must be of size k
      * @see <a href="https://en.wikipedia.org/wiki/Multinomial_distribution">Multinomial Distribution</a>
      * Generalisation of the Binomial distribution to variables with more than 2 possible values
      */
-    private Multinomial(IntegerTensor n, DoubleTensor p) {
-        checkTensorsMatchNonLengthOneShapeOrAreLengthOne(n.getShape(), TensorShape.removeDimension(0, p.getShape()));
-        Preconditions.checkArgument(
-            p.sum(0).elementwiseEquals(DoubleTensor.ones(n.getShape())).allTrue(),
-            "Probabilities must sum to one"
-        );
-        CATEGORY_PROBABILITIES_CANNOT_BE_ZERO.validate(p);
-
-        numCategories = p.getShape()[0];
+    private Multinomial(IntegerTensor n, DoubleTensor p, boolean validationEnabled) {
+        k = Ints.checkedCast(p.getShape()[p.getRank() - 1]);
         this.n = n;
         this.p = p;
-    }
+        this.validationEnabled = validationEnabled;
```

Comment:
I liked the old way, with the `DebugTensorValidator`. You could implement all these validation checks with `TensorValidator`s.

Reply:
The previous TensorValidator only did a check on the sum of the probabilities. That meant that you could turn off some of the checks but you couldn't turn them off wholesale. Converting all of these checks to TensorValidator could be done but I think delegating the checks to them doesn't make it any clearer nor add any new/needed feature. We would have to write TensorValidators for tensor shape checking which *might* make the shape validation code few lines but I don't think it's worth the extra complexity and extra classes required.